src/ws-client.cpp

```diff
@@ -0,0 +1,244 @@
+            return FALSE;
+
+        source.pfd.revents = 0;
+        return TRUE;
+    },
+    // finalize
+    [](GSource *base)
+    {
+        auto& source = *reinterpret_cast<TargetSource*>(base);
+
+        if (source.isReading) {
+            wl_display_cancel_read(source.display);
+            source.isReading = false;
+        }
+    },
+    nullptr, // closure_callback
+    nullptr, // closure_marshall
+};
+
+
+BaseBackend::BaseBackend(int hostFD)
+{
+    m_wl.display = wl_display_connect_to_fd(hostFD);
+}
+
+BaseBackend::~BaseBackend()
+{
+    if (m_wl.display)
+        wl_display_disconnect(m_wl.display);
+    m_wl.display = nullptr;
+}
+
+
+BaseTarget::BaseTarget(int hostFD, Impl& impl)
+    : m_impl(impl)
+{
+    m_glib.socket = FdoIPC::Connection::create(hostFD);
+}
+
+BaseTarget::~BaseTarget()
+{
+    if (m_wl.wpeBridgeId && m_glib.socket)
+        m_glib.socket->send(FdoIPC::Messages::UnregisterSurface, m_wl.wpeBridgeId);
+
+    g_clear_pointer(&m_wl.frameCallback, wl_callback_destroy);
+    g_clear_pointer(&m_wl.surface, wl_surface_destroy);
+
+    g_clear_pointer(&m_wl.wpeBridge, wpe_bridge_destroy);
+    g_clear_pointer(&m_wl.compositor, wl_compositor_destroy);
+    g_clear_pointer(&m_wl.eventQueue, wl_event_queue_destroy);
+
+    if (m_glib.wlSource) {
+        g_source_destroy(m_glib.wlSource);
+        g_source_unref(m_glib.wlSource);
+    }
+}
+
+void BaseTarget::initialize(struct wl_display* display)
+{
+    m_wl.eventQueue = wl_display_create_queue(display);
+
+    struct wl_registry* registry = wl_display_get_registry(display);
+    wl_proxy_set_queue(reinterpret_cast<struct wl_proxy*>(registry), m_wl.eventQueue);
+    wl_registry_add_listener(registry, &s_registryListener, this);
+    wl_display_roundtrip_queue(display, m_wl.eventQueue);
+    wl_registry_destroy(registry);
+
+    m_wl.surface = wl_compositor_create_surface(m_wl.compositor);
+    wl_proxy_set_queue(reinterpret_cast<struct wl_proxy*>(m_wl.surface), m_wl.eventQueue);
+
+    m_glib.wlSource = g_source_new(&TargetSource::s_sourceFuncs, sizeof(TargetSource));
+    {
+        auto& source = *reinterpret_cast<TargetSource*>(m_glib.wlSource);
+        source.pfd.fd = wl_display_get_fd(display);
+        source.pfd.events = G_IO_IN | G_IO_ERR | G_IO_HUP;
+        source.pfd.revents = 0;
+        source.display = display;
+        source.queue = m_wl.eventQueue;
+        source.isReading = false;
+
+        g_source_add_poll(m_glib.wlSource, &source.pfd);
+        g_source_set_name(m_glib.wlSource, "WPEBackend-fdo::wayland");
+        g_source_set_can_recurse(m_glib.wlSource, TRUE);
+        g_source_attach(m_glib.wlSource, g_main_context_get_thread_default());
+    }
+
+    wpe_bridge_add_listener(m_wl.wpeBridge, &s_bridgeListener, this);
+    wpe_bridge_connect(m_wl.wpeBridge, m_wl.surface);
+    wl_display_roundtrip_queue(display, m_wl.eventQueue);
+}
+
+void BaseTarget::requestFrame()
+{
+    if (m_wl.frameCallback)
+        std::abort();
```

Comment:
In a similar vein, `g_error()` would be nicer as it would also print a message, but this PR is only moving code around, so let's use #68 to track doing this later on.